app/models/course/material/folder.rb

```diff
@@ -21,6 +23,24 @@ def self.after_course_create(course)
     course.material_folders.create(name: 'Root')
   end
 
+  # Find the next valid name whose format is based on the the name given.
+  # If the base_name has been taken, the next name will be generate in the format of: base_name(0),
+  # base_name(1), etc.
+  # @param [String] base_name The name which we want to name the folder.
+  # @return [String] The next valid name that haven't been taken.
+  def next_valid_name(base_name = name)
+    return base_name unless parent && base_name
+
+    names_taken = existing_names.map(&:downcase)
+    name_generator = FileName.new(base_name, path: :relative, add: :always,
```

Comment:
why still n queries ? 
This gem is quite powerful for generating filenames. it can also generate numbers before the extension, like: base_name(0).txt, base_name(1).txt